pkg/services/authn/clients/ext_jwt.go

```diff
@@ -0,0 +1,251 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-jose/go-jose/v3"
+	"github.com/go-jose/go-jose/v3/jwt"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/signingkeys"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+var _ authn.Client = new(ExtendedJWT)
+
+var (
+	timeNow = time.Now
+)
+
+const (
+	SigningMethodNone     = jose.SignatureAlgorithm("none")
+	rfc9068ShortMediaType = "at+jwt"
+	rfc9068MediaType      = "application/at+jwt"
+
+	defaultOrgID = 1
+)
+
+func ProvideExtendedJWT(userService user.Service, cfg *setting.Cfg, signingKeys signingkeys.Service) *ExtendedJWT {
+	return &ExtendedJWT{
+		cfg:         cfg,
+		log:         log.New(authn.ClientExtendedJWT),
+		userService: userService,
+		signingKeys: signingKeys,
+	}
+}
+
+type ExtendedJWT struct {
+	cfg         *setting.Cfg
+	log         log.Logger
+	userService user.Service
+	signingKeys signingkeys.Service
+}
+
+func (s *ExtendedJWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identity, error) {
+	jwtToken := s.retrieveToken(r.HTTPRequest)
+
+	claims, err := s.verifyRFC9068Token(ctx, jwtToken)
+	if err != nil {
+		s.log.Error("Failed to verify JWT", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to verify JWT: %w", err)
+	}
+
+	// user:id:18
+	userID, err := strconv.ParseInt(strings.TrimPrefix(claims["sub"].(string), fmt.Sprintf("%s:id:", authn.NamespaceUser)), 10, 64)
+	if err != nil {
+		s.log.Error("Failed to parse sub", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to parse sub: %w", err)
+	}
+
+	signedInUser, err := s.userService.GetSignedInUserWithCacheCtx(ctx, &user.GetSignedInUserQuery{OrgID: r.OrgID, UserID: userID})
```

Comment:
Is it an issue if the user does not exist in the DB? 

- What if it's the first login?
- Does this work with service account JWTs?

Reply:
> What do you mean? Does the user exist or are we talking about a sign up scenario?
Sign up scenario yes

> service account JWTs

JWTs generated by the oauth2 server whose identity is linked to a service account. I think `GetSignedInUserWithCacheCtx` will return err on those if it's filtering `is_service_account`

Reply:
> > What do you mean? Does the user exist or are we talking about a sign up scenario?

> Sign up scenario yes?

This client only verifies the JWT tokens, and the token itself will be generated by the OAuth2 server which only returns a valid access token if there is a valid user in the `assertion`. So the sign up scenario is not supported by this client.

> > service account JWTs

> JWTs generated by the oauth2 server whose identity is linked to a service account. I think GetSignedInUserWithCacheCtx will return err on those if it's filtering is_service_account

With the `client_credentials` flow external services can fetch an access token whose identity is linked to a service account and it should work properly with this client. (I tested it)